mlflow/utils/file_utils.py

```diff
@@ -43,11 +44,69 @@
 from mlflow.utils import download_cloud_file_chunk
 from mlflow.utils.request_utils import download_chunk
 
-
 ENCODING = "utf-8"
 MAX_PARALLEL_DOWNLOAD_WORKERS = os.cpu_count() * 2
 
 
+class ArtifactProgressBar:
+    def __init__(self, iterable, desc, total, step, **kwargs) -> None:
+        self.iterable = iterable
+        self.total = total
+        self.step = step
+        self.pbar = None
+        self.update_pbar = False
+
+        if MLFLOW_ENABLE_ARTIFACTS_PROGRESS_BAR.get():
+            try:
+                from tqdm.auto import tqdm
+
+                self.pbar = tqdm(total=self.total, desc=desc, **kwargs)
+            except ImportError:
+                pass
+
+    @classmethod
+    def file(cls, iterable, file_size, desc, chunk_size):
+        return cls(
+            iterable,
+            desc,
+            total=file_size,
+            step=chunk_size,
+            unit="iB",
+            unit_scale=True,
+            unit_divisor=1024,
+            miniters=1,
+        )
+
+    @classmethod
+    def files(cls, iterable, desc, total, step=1):
+        return cls(iterable, desc, total=total, step=step)
+
+    def update(self):
+        self.update_pbar = True
+
+    def _pbar_iter(self):
+        for index, item in enumerate(self.iterable):
+            yield item
+            if self.update_pbar:
+                remaining = self.total - index * self.step
+                self.pbar.update(min(remaining, self.step))
+                self.pbar.refresh()
+                self.update_pbar = False
+
+    def __iter__(self):
+        if self.pbar is None:
+            yield from self.iterable
+        else:
+            yield from self._pbar_iter()
```

Comment:
I think we can remove this. This class no longer needs to work as an iterator.

Reply:
I see :) Updated